_posts/2020-09-06-how-the-system-tests-works.md

```diff
@@ -0,0 +1,603 @@
+Note that pointer stack is set automatically in `AbstractST.class` in `@BeforeAll` or `@BeforeEach` methods.
+
+Cluster Operator setup example:
+```java
+    @BeforeAll
+    void createClassResources() {
+        prepareEnvForOperator(NAMESPACE);                          <--- Create namespaces
+        createTestClassResources();                                <--- Create Resources instance for class
+        applyRoleBindings(NAMESPACE);                              <--- Apply Cluster Operator bindings
+        KubernetesResource.clusterOperator(NAMESPACE).done();      <--- Deploy Cluster Operator
+        ...
+    }
+```
+
+#### Exercise
+In this phase we specify all steps which need to be executed to cover some specific functionality. Reasonable example would be
+the Code snippet 1.4 from previous section, where we setup Kafka cluster with new persistent claim storage.
+
+#### Verify
+
+When your environment is successfully deployed and settled from the previous phase, you can add code for some checks, message exchange, etc.
+
+#### Teardown
+
+Because we have two stacks for storing resources, cluster resources deletion can be easily performed in `@AfterEach` or `@AfterAll` methods.
+Resource lifecycle implementation will ensure that all resources tied to a specific stack will be deleted in the correct order.
+Teardown is triggered in `@AfterAll` of `AbstractST`:
+```java
+    @AfterAll
+    void teardownEnvironmentClass() {
+        if (Environment.SKIP_TEARDOWN == null) {
+            tearDownEnvironmentAfterAll();
+            teardownEnvForOperator();
+        }
+    }
+```
+
+so if you want to change teardown from your `@AfterAll`, you must override method `tearDownEnvironmentAfterAll()`:
+```java
+    @Override
+    protected void tearDownEnvironmentAfterAll() {
+        doSomethingYouNeed();
+        super.tearDownEnvironmentAfterAll();
+    }
+```
+
+In order to delete all resources from specific `Resources` instance, execute:
+```java
+    ResourceManager.deleteMethodResources();
+    ResourceManager.deleteClassResources();
+```
+
+## Auxiliary classes
+
+In `system test` module there are many helper classes,  each having a different purpose:
+
+1. Utils
+2. Constants
+3. Environment
+4. Kubernetes client
+5. Kafka clients
+
+#### Utils
+
+These classes are applicable on various problems. We have separate utils class for each resource. 
+For instance we have KafkaUtils for KafkaResource, KafkaTopic for KafkaResource and so on. Furthermore, we have got
+Kubernetes related utils such as ServiceUtils, PodUtils, DeploymentsUtils and many more. The hierarchy is as follows:
+
+* utils
+  * kafkaUtils
+    * KafkaBridgeUtils
+    * KafkaConnectS2IUtils
+    * ...
+  * kubeUtils
+    * controllers
+      * DeploymentConfigUtils
+      * DeploymentUtils
+      * ...
+    * objects
+      * NamespaceUtils
+      * PersistentVolumeClaimUtils
+      * ...
+  * specific
+    * BridgeUtils
+    * CruiseControlUtils
```

Comment:
What is this structure? packages and classes? Is it really needed when you anyway, always end up with `...`?

Reply:
removed